src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp

```diff
@@ -412,7 +412,113 @@ bool PDPServer2::process_change_acknowledgement(
     return true;
 }
 
-fastdds::rtps::ddb::DiscoveryDataBase& PDPServer2:: discovery_db()
+bool PDPServer2::process_disposals()
+{
+    EDPServer2* edp = static_cast<EDPServer2*>(mp_EDP);
+    fastrtps::rtps::WriterHistory* pubs_history = edp->publications_writer_.second;
+    fastrtps::rtps::WriterHistory* subs_history = edp->subscriptions_writer_.second;
+
+    // Get list of disposals from database
+    std::vector<fastrtps::rtps::CacheChange_t*> disposals = discovery_db_.changes_to_dispose();
+    // Iterate over disposals
+    for(auto change: disposals)
+    {
+        // Check that the change is actually a disposal
+        if (change->kind != fastrtps::rtps::ChangeKind_t::NOT_ALIVE_DISPOSED)
+        {
+            logWarning(PDPServer2, "The change does not correspond with a disposal");
+            continue;
+        }
+
+        fastrtps::rtps::GuidPrefix_t change_guid_prefix = discovery_db_.guid_from_change(change).guidPrefix;
+
+        // DATA(Up) case
+        if (discovery_db_.is_participant(change))
+        {
+            // Lock PDP writer
+            std::unique_lock<fastrtps::RecursiveTimedMutex> lock(mp_PDPWriter->getMutex());
+
+            // Remove all DATA(p) with the same sample identity as the DATA(Up) from PDP writer's history.
+            remove_related_alive_from_history_nts(mp_PDPWriterHistory, change_guid_prefix);
+
```

Comment:
This is more a design issue than your code:
When a Data(Up) arrives, it should check if there is any Data(w) or Data(r) alreay in the history from him.
With this change you can skip completly the Data(Uw) and Data(Ur) steps in case of a Participant in disposals, and skip the mutex take and to check the history

Reply:
In agreement with the reviewer, we have decided to leave the algorithm as-is for now